metronome/hotstuff/consensus/src/metronome/hotstuff/consensus/basic/ProtocolState.scala

```diff
@@ -0,0 +1,492 @@
+    * lower level protocol message; we know the signatures are correct;
+    * and the contents of any proposed block have been validated as well,
+    * so they are safe to be voted on.
+    *
+    * Return the updated state and any effects to be carried out in response,
+    * or an error, so that mismatches can be traced. Discrepancies can arise
+    * from the state being different or have changed since the message originally
+    * received.
+    *
+    * The structure of the method tries to match the pseudo code of `Algorithm 2`
+    * in the HotStuff paper.
+    */
+  def handleMessage(
+      e: Validated[MessageReceived[A]]
+  ): TransitionAttempt[A] =
+    phase match {
+      // Leader:  Collect NewViews, create block, boradcast Prepare
+      // Replica: Wait for Prepare, check safe extension, vote Prepare, move to PreCommit.
+      case Phase.Prepare =>
+        matchingMsg(e) {
+          case m: NewView[_] if m.viewNumber == viewNumber.prev && isLeader =>
+            Right(addNewViewAndMaybeCreateBlock(e.sender, m))
+
+          case m: Prepare[_] if matchingLeader(e) =>
+            if (isSafe(m)) {
+              val blockHash = Block[A].blockHash(m.block)
+              val effects = Seq(
+                sendVote(Phase.Prepare, blockHash)
+              )
+              val next = moveTo(Phase.PreCommit).copy(
+                preparedBlockHash = blockHash
+              )
+              Right(next -> effects)
+            } else {
+              Left(UnsafeExtension(e.sender, m))
+            }
+        }
+
+      // Leader:  Collect Prepare votes, broadcast Prepare Q.C.
+      // Replica: Wait for Prepare Q.C, save prepareQC, vote PreCommit, move to Commit.
+      case Phase.PreCommit =>
+        matchingMsg(e) {
+          handleVotes(e, Phase.Prepare) orElse
+            handleQuorum(e, Phase.Prepare) { m =>
+              val effects = Seq(
+                sendVote(Phase.PreCommit, m.quorumCertificate.blockHash)
+              )
+              val next = moveTo(Phase.Commit).copy(
+                prepareQC = m.quorumCertificate
+              )
+              next -> effects
+            }
+        }
+
+      // Leader:  Collect PreCommit votes, broadcast PreCommit Q.C.
+      // Replica: Wait for PreCommit Q.C., save lockedQC, vote Commit, move to Decide.
+      case Phase.Commit =>
+        matchingMsg(e) {
+          handleVotes(e, Phase.PreCommit) orElse
+            handleQuorum(e, Phase.PreCommit) { m =>
+              val effects = Seq(
+                sendVote(Phase.Commit, m.quorumCertificate.blockHash)
+              )
+              val next = moveTo(Phase.Decide).copy(
+                lockedQC = m.quorumCertificate
+              )
+              next -> effects
+            }
+        }
+
+      // Leader:  Collect Commit votes, broadcast Commit Q.C.
+      // Replica: Wait for Commit Q.C., execute block, send NewView, move to Prepare.
+      case Phase.Decide =>
```

Comment:
I think since we are not doing Chained Hotstuff, during the decision phase, we should clear out variables like lockedQC, etc. used in the previous view.

Reply:
So you're saying that if we reached the decide phase in view N, then the `lockedQC` of view `N` should be forgotten? And by "like" you mean `prepareQC` as well?

There's no other "previous view", there is just 1 `lockedQC`. If we clear it out, then in view `N+1` we can't use it in the `isSafe` method to check that the block proposed by the next leader extends it. We also cannot clear out `prepareQC` because if `N+1` times out, then it's the only `prepareQC` we can send to the leader of `N+2`.

Or did I misunderstand your suggestion?

Reply:
Yeah, I misunderstood the protocol and there is no need to clear out both QCs in this case.